### PR TITLE
Add integration with lgtm static analysis

### PR DESCRIPTION
**Pull request by Smjert**

### CODE REVIEW

**Comment by theopolis** [comment]
Very cool!

**Comment by Smjert** [comment]
> I know the work is ongoing. But I'd be fine merging this as is, and iterating. It's clearly further along than before

Well to be fair I'm still trying to understand if it's feasible to have it run at each PR at all.
The C++ build/analysis timed out after 4 hours; the build took 2h 40m.
Now I've disabled building the tests although I suspect it might hide some issues.

**Comment by directionless** [comment]
> if it's feasible to have it run at each PR at all

Even if we don't run it on the PRs. I think it's still worth having the config file.

I also think that we regularly let PRs sit for days. So there may be value in having it run, but not be required. (As it is today)

**Comment by Smjert** [comment]
> > if it's feasible to have it run at each PR at all
> 
> Even if we don't run it on the PRs. I think it's still worth having the config file.
> 
> I also think that we regularly let PRs sit for days. So there may be value in having it run, but not be required. (As it is today)

Well, just didn't want to be a reason to wait on merging a PR though.
That been said, even with all the optimizations to the build speed I can think of, it's still taking more than 4h total, which is the maximum limit.

**Comment by mike-myers-tob** [comment]
> That been said, even with all the optimizations to the build speed I can think of, it's still taking more than 4h total, which is the maximum limit.

Is it possible to configure it to ignore certain subdirectories? Specifically I'm thinking we could ignore [osquery/third-party](https://github.com/osquery/osquery/tree/master/third-party)

File: .lgtm.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+path_classifiers:
```
Comment (directionless) [inline .lgtm.yml]:
```suggestion
# https://lgtm.com/help/lgtm/lgtm.yml-configuration-file
path_classifiers:
```

File: .lgtm.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  library:
+    exclude: "libraries/**/*"
+
```
Comment (directionless) [inline .lgtm.yml]:
```suggestion
  third-party:
    exclude: "third-party/**/*"

```

maybe?

**Comment by Smjert** [comment]
> > That been said, even with all the optimizations to the build speed I can think of, it's still taking more than 4h total, which is the maximum limit.
> 
> Is it possible to configure it to ignore certain subdirectories? Specifically I'm thinking we could ignore [osquery/third-party](https://github.com/osquery/osquery/tree/master/third-party)

There's a problem at the "base", which is that it takes 2h to build:
```
[2020-02-13 12:55:14] [build] + cmake --build . -j 4

[...]

[2020-02-13 14:57:20] [build] [100%] Linking CXX executable osqueryd
[2020-02-13 14:57:26] [build] [100%] Built target osqueryd
[2020-02-13 14:57:26] [build] Scanning dependencies of target create_osqueryi
[2020-02-13 14:57:26] [build] [100%] Generating osqueryi
[2020-02-13 14:57:26] [build] [100%] Built target create_osqueryi
```


@directionless lgtm works by tracking the sources that the compiler touches to build, so given that `third-party` doesn't contain code, marking it as library won't improve the analysis speed.

In any case, by the time it starts the analysis it's almost 2h30m in, and then the 150 checks it does consume all the rest. 
This time it stopped at:
```
[2020-02-13 16:21:45] [analysis] [EVALUATION 136/150] Completed query semmlecode-cpp-queries/Metrics/Dependencies/ExternalDependenciesSourceLinks.ql (0s)
```

I guess we could try to trim down the queries, although that's not the main build time driver.

**Comment by theopolis** [comment]
How is this going?

**Comment by Smjert** [comment]
> How is this going?

This is a bit on my backlog unfortunately; this depends on https://github.com/osquery/osquery/pull/6465 and https://github.com/osquery/osquery/pull/6464.
Then we need to refactor a bit how we deal with third part libraries, to be able to link to system libraries.

**Comment by directionless** [comment]
I think this is still worth merging. it's better than today

**Comment by gbedoya** [comment]
Also, I've noticed that LGTM C/C++ analysis fails as a result of timing out.

For example ...
https://lgtm.com/projects/g/osquery/osquery/logs/rev/pr-68f111dbebf2221b4fc7a37948b25706f4668f3a/lang:cpp/stage:Build%20master_34cc6f141159a2601b0c9ec248cb23c5a41efd76

Notice the execution time of best practice checks ...
```
[2020-05-31 03:01:11] [analysis] [EVALUATION 7/150] Completed query semmlecode-cpp-queries/Best Practices/UseOfGoto.ql (291s)
[2020-05-31 03:01:16] [analysis] [EVALUATION 8/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedStaticVariables.ql (297s)
[2020-05-31 03:01:48] [analysis] [EVALUATION 9/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedStaticFunctions.ql (328s)
[2020-05-31 03:01:56] [analysis] [EVALUATION 10/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedLocals.ql (337s)
[2020-05-31 03:02:24] [analysis] [EVALUATION 11/150] Completed query semmlecode-cpp-queries/Best Practices/SwitchLongCase.ql (364s)
[2020-05-31 03:02:24] [analysis] [EVALUATION 12/150] Completed query semmlecode-cpp-queries/Best Practices/SloppyGlobal.ql (364s)
[2020-05-31 03:03:34] [analysis] [EVALUATION 13/150] Completed query semmlecode-cpp-queries/Best Practices/RuleOfTwo.ql (435s)
[2020-05-31 03:03:44] [analysis] [EVALUATION 14/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/Slicing.ql (445s)
[2020-05-31 03:04:12] [analysis] [EVALUATION 15/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/OffsetUseBeforeRangeCheck.ql (473s)
[2020-05-31 03:05:11] [analysis] [EVALUATION 16/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/EmptyBlock.ql (531s)
[2020-05-31 03:05:30] [analysis] [EVALUATION 17/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/LocalVariableHidesGlobalVariable.ql (551s)
[2020-05-31 03:05:52] [analysis] [EVALUATION 18/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/DeclarationHidesVariable.ql (572s)
[2020-05-31 03:06:16] [analysis] [EVALUATION 19/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/DeclarationHidesParameter.ql (596s)
[2020-05-31 03:06:18] [analysis] [EVALUATION 20/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/ThrowingPointers.ql (599s)
[2020-05-31 03:06:44] [analysis] [EVALUATION 21/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/LeakyCatch.ql (625s)
[2020-05-31 03:07:38] [analysis] [EVALUATION 24/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/AccidentalRethrow.ql (679s)
[2020-05-31 03:07:46] [analysis] [EVALUATION 27/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/CatchingByValue.ql (687s)
[2020-05-31 03:07:50] [analysis] [EVALUATION 28/150] Completed query semmlecode-cpp-queries/Best Practices/ComplexCondition.ql (691s)
[2020-05-31 03:07:59] [analysis] [EVALUATION 29/150] Completed query semmlecode-cpp-queries/Best Practices/BlockWithTooManyStatements.ql (700s)
```

The sum is 9565 seconds (or ~2h39mins). Further notice there's 150 C/C++ checks. In this LGTM analysis 126 checks ran before timing out. Perhaps, we can exclude best practice checks?

**Comment by Smjert** [comment]
> Also, I've noticed that LGTM C/C++ analysis fails as a result of timing out.
> 
> For example ...
> https://lgtm.com/projects/g/osquery/osquery/logs/rev/pr-68f111dbebf2221b4fc7a37948b25706f4668f3a/lang:cpp/stage:Build%20master_34cc6f141159a2601b0c9ec248cb23c5a41efd76
> 
> Notice the execution time of best practice checks ...
> 
> ```
> [2020-05-31 03:01:11] [analysis] [EVALUATION 7/150] Completed query semmlecode-cpp-queries/Best Practices/UseOfGoto.ql (291s)
> [2020-05-31 03:01:16] [analysis] [EVALUATION 8/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedStaticVariables.ql (297s)
> [2020-05-31 03:01:48] [analysis] [EVALUATION 9/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedStaticFunctions.ql (328s)
> [2020-05-31 03:01:56] [analysis] [EVALUATION 10/150] Completed query semmlecode-cpp-queries/Best Practices/Unused Entities/UnusedLocals.ql (337s)
> [2020-05-31 03:02:24] [analysis] [EVALUATION 11/150] Completed query semmlecode-cpp-queries/Best Practices/SwitchLongCase.ql (364s)
> [2020-05-31 03:02:24] [analysis] [EVALUATION 12/150] Completed query semmlecode-cpp-queries/Best Practices/SloppyGlobal.ql (364s)
> [2020-05-31 03:03:34] [analysis] [EVALUATION 13/150] Completed query semmlecode-cpp-queries/Best Practices/RuleOfTwo.ql (435s)
> [2020-05-31 03:03:44] [analysis] [EVALUATION 14/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/Slicing.ql (445s)
> [2020-05-31 03:04:12] [analysis] [EVALUATION 15/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/OffsetUseBeforeRangeCheck.ql (473s)
> [2020-05-31 03:05:11] [analysis] [EVALUATION 16/150] Completed query semmlecode-cpp-queries/Best Practices/Likely Errors/EmptyBlock.ql (531s)
> [2020-05-31 03:05:30] [analysis] [EVALUATION 17/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/LocalVariableHidesGlobalVariable.ql (551s)
> [2020-05-31 03:05:52] [analysis] [EVALUATION 18/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/DeclarationHidesVariable.ql (572s)
> [2020-05-31 03:06:16] [analysis] [EVALUATION 19/150] Completed query semmlecode-cpp-queries/Best Practices/Hiding/DeclarationHidesParameter.ql (596s)
> [2020-05-31 03:06:18] [analysis] [EVALUATION 20/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/ThrowingPointers.ql (599s)
> [2020-05-31 03:06:44] [analysis] [EVALUATION 21/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/LeakyCatch.ql (625s)
> [2020-05-31 03:07:38] [analysis] [EVALUATION 24/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/AccidentalRethrow.ql (679s)
> [2020-05-31 03:07:46] [analysis] [EVALUATION 27/150] Completed query semmlecode-cpp-queries/Best Practices/Exceptions/CatchingByValue.ql (687s)
> [2020-05-31 03:07:50] [analysis] [EVALUATION 28/150] Completed query semmlecode-cpp-queries/Best Practices/ComplexCondition.ql (691s)
> [2020-05-31 03:07:59] [analysis] [EVALUATION 29/150] Completed query semmlecode-cpp-queries/Best Practices/BlockWithTooManyStatements.ql (700s)
> ```
> 
> The sum is 9565 seconds (or ~2h39mins). Further notice there's 150 C/C++ checks. In this LGTM analysis 126 checks ran before timing out. Perhaps, we can exclude best practice checks?

I've already disabled the "recommended" category, because they originally where almost double. Some of those best-practices ones are definitely wanted.
There are many runs where it is only able to run ~40 queries, so it's not really a problem of the analysis phase, what's taking all the time is the build.

**Comment by theopolis** [comment]
We discussed this in office hours today and decided to pause the LGTM integration and instead wait to use the new GitHub integrations. https://github.com/osquery/foundation/pull/66

**Comment by mike-myers-tob** [comment]
LGTM.com is being discontinued in December 2022, and people are being referred to https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/setting-up-code-scanning-for-a-repository#setting-up-code-scanning-manually

**Comment by Smjert** [comment]
> LGTM.com is being discontinued in December 2022, and people are being referred to https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/setting-up-code-scanning-for-a-repository#setting-up-code-scanning-manually

Yeah I've an implementation of that on my repo done some time ago; I've been able to trim the compilation enough to make the analysis work in reasonable time. The remaining problem is that while I can direct the analysis to skip third party libraries code, it still picks up issues in their headers, since osquery code includes them, and there's no easy way to filter it like in other languages they support.
The only way right now would be to create our own queries which filter out the third party libraries paths.

**Pull request closed without merging**